pkg/sql/grant_revoke_test.go

```diff
@@ -0,0 +1,223 @@
+			require.True(t, ok)
+			require.True(t, priv.IsSetIn(userPriv.Privileges))
+			require.Equal(t, descVersion, desc.GetVersion())
+		}
+	}
+}
+
+// TestNoOpRevoke tests that if a REVOKE privilege statement is a no-op
+// (e.g. REVOKE SELECT on a table that a user does not have SELECT privilege on)
+// then the statement is actually a no-op (i.e. no schema change will happen).
+func TestNoOpRevoke(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	params, _ := tests.CreateTestServerParams()
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop(context.Background())
+	tdb := sqlutils.MakeSQLRunner(sqlDB)
+
+	tdb.Exec(t, "CREATE DATABASE db")
+	tdb.Exec(t, "CREATE SCHEMA db.sc")
+	tdb.Exec(t, "CREATE TABLE db.sc.tbl(a int)")
+	tdb.Exec(t, "CREATE TYPE db.sc.typ AS ENUM('a', 'b')")
+	tdb.Exec(t, "CREATE USER roach")
+
+	// Assert that user `roach` should not have any privilege on the database, schema, and table.
+	dbDesc := desctestutils.TestingGetDatabaseDescriptor(kvDB, keys.SystemSQLCodec, "db")
+	scDesc := desctestutils.TestingGetSchemaDescriptor(kvDB, keys.SystemSQLCodec, dbDesc.GetID(), "sc")
+	tblDesc := desctestutils.TestingGetTableDescriptor(kvDB, keys.SystemSQLCodec, "db", "sc", "tbl")
+	typDesc := desctestutils.TestingGetTypeDescriptor(kvDB, keys.SystemSQLCodec, "db", "sc", "typ")
+	userRoach, err := username.MakeSQLUsernameFromUserInput("roach", username.PurposeValidation)
+	require.NoError(t, err)
+	_, ok := dbDesc.GetPrivileges().FindUser(userRoach)
+	require.False(t, ok)
+	_, ok = scDesc.GetPrivileges().FindUser(userRoach)
+	require.False(t, ok)
+	_, ok = tblDesc.GetPrivileges().FindUser(userRoach)
+	require.False(t, ok)
+	_, ok = typDesc.GetPrivileges().FindUser(userRoach)
+	require.False(t, ok)
+
+	retrieveDescriptorByObjectType := func(objectType privilege.ObjectType) catalog.Descriptor {
+		var desc catalog.Descriptor
+		switch objectType {
+		case privilege.Database:
+			desc = desctestutils.TestingGetDatabaseDescriptor(kvDB, keys.SystemSQLCodec, dbDesc.GetName())
+		case privilege.Schema:
+			desc = desctestutils.TestingGetSchemaDescriptor(kvDB, keys.SystemSQLCodec, dbDesc.GetID(), scDesc.GetName())
+		case privilege.Table:
+			desc = desctestutils.TestingGetTableDescriptor(kvDB, keys.SystemSQLCodec, dbDesc.GetName(), scDesc.GetName(), tblDesc.GetName())
+		case privilege.Type:
+			desc = desctestutils.TestingGetTypeDescriptor(kvDB, keys.SystemSQLCodec, dbDesc.GetName(), scDesc.GetName(), typDesc.GetName())
+		}
+		return desc
+	}
+
+	testCases := []struct {
+		objectType privilege.ObjectType
+		// privsToTest will contain all allowed privileges for this object type except for 'ALL' and 'GRANT'
+		privsToTest  []privilege.Kind
+		objectName   string
+		retrieveDesc func(...interface{}) catalog.Descriptor
+	}{
+		{
+			objectType:  privilege.Database,
+			privsToTest: []privilege.Kind{privilege.CONNECT, privilege.CREATE, privilege.DROP, privilege.ZONECONFIG},
```

Comment:
As commented above, at least let's try to get the privileges for the kind programmatically such that if or when we added new ones, it would show up here and so that there's no guessing